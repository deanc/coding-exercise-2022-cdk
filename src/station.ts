import { Point } from "./point";

/**
 * Represents a link station
 * @param coordinates the x,y co-ordinates
 * @param reach       the reach of the station
 */
export class Station {
  readonly coordinates: Point;
  readonly reach: number;
  power: number;

  constructor(coordinates: Point, reach: number) {
    this.coordinates = coordinates;
    this.reach = reach;
    this.power = 0;
  }

  /**
   * Gets the coordinates of the link station
   * @return the point
   */
  getCoordinates(): Point {
    return this.coordinates;
  }

  /**
   * Gets the individual x-cordinate of the station
   * @return x
   */
  getX(): number {
    return this.coordinates[0];
  }

  /**
   * Gets the individual y-cordinate of the station
   * @return y
   */
  getY(): number {
    return this.coordinates[1];
  }

  /**
   * Gets the reach of the station
   * @return the reach
   */
  getReach(): number {
    return this.reach;
  }

  /**
   * Gets the power of the station
   * @return power
   */
  getPower(): number {
    return this.power;
  }

  /**
   * Sets the power of the station
   */
  setPower(power: number): void {
    this.power = power;
  }
}
