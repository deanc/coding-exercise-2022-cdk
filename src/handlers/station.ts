import { APIGatewayProxyEventV2, APIGatewayProxyResultV2 } from 'aws-lambda';

import { Point } from '../point';
import { Station } from '../station';
import { findBestStation } from '../station_finder';

export async function main(
    event: APIGatewayProxyEventV2
): Promise<APIGatewayProxyResultV2> {
    console.log('request:', JSON.stringify(event, undefined, 2));

    if (!event.body) {
        return errorHandler('Empty POST body');
    }

    const parsedBody = JSON.parse(event.body);

    // NOTE: we would validate the JSON object here in a real-world scenario before
    // casting it to our desired type. For now we have basic validation below.
    const body = parsedBody as StationFinderRequest;

    const { stations, points } = body;
    if (!body.stations || !body.points) {
        return errorHandler('Invalid parameters');
    }

    const mappedStations = stationsMapper(stations);

    const response = points.map((point: Point) => {
        const bestStation = findBestStation(mappedStations, point);

        return {
            point,
            ...(bestStation && {
                station: {
                    x: bestStation.getX(),
                    y: bestStation.getY(),
                    power: bestStation.getPower(),
                },
            }),
        };
    });

    return {
        statusCode: 200,
        headers: {},
        body: JSON.stringify(response),
    };
}

const errorHandler = (
    message: string,
    statusCode = 400
): APIGatewayProxyResultV2 => {
    return {
        statusCode,
        headers: {},
        body: JSON.stringify({ message }),
    };
};

interface IncomingStation {
    coordinates: Point;
    reach: number;
}

interface StationFinderRequest {
    points: Point[];
    stations: IncomingStation[];
}

const stationsMapper = (stationObjects: IncomingStation[]): Station[] =>
    stationObjects.map((obj) => new Station(obj.coordinates, obj.reach));
