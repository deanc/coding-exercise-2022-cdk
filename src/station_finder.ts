import { getDistance, getPower } from './misc';
import { Point } from './point';
import { Station } from './station';

/**
 * Utility function to remove stations whose power is not enough
 * Should be passed into Array.filter()
 * @param  station Station
 * @return
 */
export const removeOutOfReachStations = (station: Station): boolean =>
    station.getPower() > 0;

/**
 * Comparison function to find the highest powered link station
 * @param  a
 * @param  b
 * @return   [description]
 */

const highestPoweredStation = (a: Station, b: Station): any =>
    a.getPower() < b.getPower();

/**
 * Finds the best link station for a given point [x,y]
 * @param  point the point to search against
 * @return       the best link station
 */
export const findBestStation = (
    stations: Station[],
    point: Point
): Station | undefined =>
    stations
        .map((station) => {
            const power = getPower(
                station.getReach(),
                getDistance(point, station.getCoordinates())
            );
            station.setPower(power);
            return station;
        })
        .filter(removeOutOfReachStations)
        .sort(highestPoweredStation)
        .pop();
