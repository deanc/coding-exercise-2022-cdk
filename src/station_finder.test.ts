import { Point } from './point';
import { Station } from './station';
import { findBestStation, removeOutOfReachStations } from './station_finder';

describe('station_finder', () => {
    it('removeOutOfReachStations removes stations out of reach', () => {
        const station = new Station([0, 0], 5);
        station.setPower(5);
        expect(removeOutOfReachStations(station)).toBe(true);
    });

    const points: Point[] = [
        [0, 0],
        [100, 100],
        [15, 10],
        [18, 18],
    ];

    describe.each(points)('finds the best station for point', (x, y) => {
        it('finds the best station', () => {
            const stations = [
                new Station([0, 0], 10),
                new Station([20, 20], 5),
                new Station([10, 0], 12),
            ];
            expect(findBestStation(stations, [x, y])).toMatchSnapshot();
        });
    });
});
