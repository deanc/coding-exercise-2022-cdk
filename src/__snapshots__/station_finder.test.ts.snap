// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`station_finder finds the best station for point finds the best station 1`] = `
Station {
  "coordinates": Array [
    10,
    0,
  ],
  "power": 4,
  "reach": 12,
}
`;

exports[`station_finder finds the best station for point finds the best station 2`] = `undefined`;

exports[`station_finder finds the best station for point finds the best station 3`] = `
Station {
  "coordinates": Array [
    10,
    0,
  ],
  "power": 0.6718427000252355,
  "reach": 12,
}
`;

exports[`station_finder finds the best station for point finds the best station 4`] = `
Station {
  "coordinates": Array [
    20,
    20,
  ],
  "power": 4.715728752538098,
  "reach": 5,
}
`;
